pkg/controller/custom_resource_replicas.go

```diff
@@ -0,0 +1,150 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	scaleclient "k8s.io/client-go/scale"
+
+	"github.com/grafana/rollout-operator/pkg/config"
+)
+
+func (c *RolloutController) adjustStatefulSetsGroupReplicasToMirrorResource(ctx context.Context, groupName string, sets []*appsv1.StatefulSet) (bool, error) {
+	// Return early no matter what after scaling up or down a single StatefulSet to make sure that rollout-operator
+	// works with up-to-date models.
+	for _, sts := range sets {
+		currentReplicas := *sts.Spec.Replicas
+
+		scaleObj, referenceGVR, referenceName, err := getCustomScaleResourceForStatefulset(ctx, sts, c.restMapper, c.scaleClient)
+		if err != nil {
+			return false, err
+		}
+		if scaleObj == nil {
+			continue
+		}
+
+		referenceResource := fmt.Sprintf("%s/%s", referenceGVR.Resource, referenceName)
+
+		desiredReplicas := scaleObj.Spec.Replicas
+		if currentReplicas == desiredReplicas {
+			updateStatusReplicasOnMirroredResourceIfNeeded(ctx, c.logger, c.dynamicClient, sts, scaleObj, referenceGVR, referenceName, desiredReplicas)
+			// No change in the number of replicas: don't log because this will be the result most of the time.
+			continue
+		}
+
+		direction := ""
+		if desiredReplicas > currentReplicas {
+			direction = "up"
+		} else if desiredReplicas < currentReplicas {
+			direction = "down"
+		}
+
+		level.Info(c.logger).Log("msg", fmt.Sprintf("scaling %s statefulset to match reference resource", direction),
+			"group", groupName,
+			"name", sts.GetName(),
+			"currentReplicas", currentReplicas,
+			"desiredReplicas", desiredReplicas,
+			"referenceResource", referenceResource,
+		)
+
+		if err := c.patchStatefulSetSpecReplicas(ctx, sts, desiredReplicas); err != nil {
+			return false, err
+		}
+
+		// Make sure that scaleObject's status.replicas field is up-to-date.
+		if scaleObj.Status.Replicas != desiredReplicas {
```

Comment:
This `if` condition is also inside `updateStatusReplicasOnMirroredResourceIfNeeded()`. I think we can either remove it from here, or remove it from `updateStatusReplicasOnMirroredResourceIfNeeded()` and add this if condition even above where we call `updateStatusReplicasOnMirroredResourceIfNeeded()`.

Reply:
Good point. I will keep the condition here to match name of the function ("if needed").